modules/migrations/migrate.go

```diff
@@ -34,39 +36,60 @@ func RegisterDownloaderFactory(factory base.DownloaderFactory) {
 	factories = append(factories, factory)
 }
 
-func isMigrateURLAllowed(remoteURL string) error {
+// IsMigrateURLAllowed checks if an URL is allowed to be migrated from
+func IsMigrateURLAllowed(remoteURL string, doer *models.User) error {
+	// Remote address can be HTTP/HTTPS/Git URL or local path.
 	u, err := url.Parse(strings.ToLower(remoteURL))
 	if err != nil {
-		return err
+		return &models.ErrInvalidCloneAddr{IsURLError: true}
 	}
 
-	if strings.EqualFold(u.Scheme, "http") || strings.EqualFold(u.Scheme, "https") {
-		if len(setting.Migrations.AllowedDomains) > 0 {
-			if !allowList.Match(u.Host) {
-				return &models.ErrMigrationNotAllowed{Host: u.Host}
-			}
-		} else {
-			if blockList.Match(u.Host) {
-				return &models.ErrMigrationNotAllowed{Host: u.Host}
-			}
+	if u.Scheme == "file" || u.Scheme == "" {
+		if !doer.CanImportLocal() {
+			return &models.ErrInvalidCloneAddr{Host: "<LOCAL_FILESYSTEM>", IsPermissionDenied: true}
+		}
+		isAbs := filepath.IsAbs(u.Host + u.Path)
+		if !isAbs {
+			return &models.ErrInvalidCloneAddr{Host: "<LOCAL_FILESYSTEM>", IsInvalidPath: true}
+		}
+		isDir, err := util.IsDir(u.Host + u.Path)
+		if err != nil {
+			log.Error("Unable to check if %s is a directory: %v", u.Host+u.Path, err)
+			return err
+		}
+		if !isDir {
+			return &models.ErrInvalidCloneAddr{Host: "<LOCAL_FILESYSTEM>", IsInvalidPath: true}
 		}
+
+		return nil
+	}
+
+	if u.Scheme == "git" && u.Port() != "" && (strings.Contains(remoteURL, "%0d") || strings.Contains(remoteURL, "%0a")) {
```

Comment:
This seems subtle, might make sense to separate this condition into its own func so it's less likely to occur again that checks for the same thing are implemented differently

Reply:
This was in its own function that wasn't being used! The point is to make sure that there really is only place where this is done.